more handling of messaging errors

snopoke opened the pull request:
## Summary
A few small changes to improve sms error handling.

See individual commits for changes.

## Product Description
On potential product change here is that messages with no actual message content will not be sent. Twilio and potentially other gateways reject messages with no content (it seems that the intent was always to not try to send messages with no content but at least one place wasn't checking correctly).

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

snopoke committed:
don't attempt to send blank messages

This generates a 400 error from Twilio

snopoke committed:
rate limit based on the specific exception

This should allow more errors to get through to Sentry while
still rate limiting when the gateway is down

snopoke committed:
handle twilio blacklist errors

snopoke committed:
handle twilio region permission error

mkangia commented on corehq/apps/sms/api.py:
```diff
@@ -368,13 +368,13 @@ def _get_backend_tag(backend=None, backend_id=None):
     """
     client = get_redis_client()
-    key = 'exception-logged-for-backend-%s' % backend.couch_id
+    key = f'exception-logged-for-backend-{backend.couch_id}-{hash(str(exception))}'
```
nit: if we are editing it anyway, should we make this key smaller.
`exception-logged-{backend.couch_id}-{hash(str(exception))}`

snopoke replied:
I don't think the length there is much of an issue since the number of keys should be quite low

mkangia replied:
Cool. Just thought it was overly descriptive.

mkangia commented on corehq/messaging/smsbackends/twilio/models.py:
```diff
@@ -16,6 +16,7 @@
 # https://www.twilio.com/docs/api/errors/reference
 INVALID_TO_PHONE_NUMBER_ERROR_CODE = 21211
 WHATSAPP_LIMITATION_ERROR_CODE = 63032
+TO_FROM_BLACKLIST_ERROR = 21610
```
nit: how about using the same naming convention ending with `_ERROR_CODE`?

snopoke replied:
cf95d1f4374

mkangia commented on corehq/messaging/smsbackends/twilio/models.py:
```diff
@@ -108,6 +109,8 @@ def send(self, msg, orig_phone_number=None, *args, **kwargs):
                 msg.set_gateway_error("Message From/To pair violates a gateway blacklist rule")
                 return
+            elif e.code == REGION_PERMISSION_ERROR:
+                msg.set_gateway_error("Destination region not enabled for this backend.")
```
do we need a "return" here?

snopoke replied:
73e64b7e326

snopoke committed:
link subevent to messages for survey sms's

This will fix a bug where searching by phone number
doesn't pick up the events in the messaging event report
since the SMS isn't linked to the messaging event

snopoke committed:
stick to naming convention

snopoke committed:
add missing return

orangejenny commented on corehq/apps/sms/models.py:
```diff
@@ -177,6 +177,7 @@ class SMSBase(UUIDGeneratorMixin, Log):
     ERROR_PHONE_NUMBER_OPTED_OUT = 'PHONE_NUMBER_OPTED_OUT'
     ERROR_INVALID_DESTINATION_NUMBER = 'INVALID_DESTINATION_NUMBER'
     ERROR_MESSAGE_TOO_LONG = 'MESSAGE_TOO_LONG'
+    ERROR_MESSAGE_BLANK = 'MESSAGE_BLANK'
```
`ERROR_NO_MESSAGE` looks like it has the same purpose: https://github.com/dimagi/commcare-hq/blob/567f7256acb30e71674f4fb1e3be66ff25dab477/corehq/messaging/scheduling/models/content.py#L48-L50

snopoke replied:
Is is the same purpose but not all messages go through this workflow, specifically sms surveys: https://github.com/dimagi/commcare-hq/blob/dd7129005ba038441a53e6a7c1951c0ce29e4d14/corehq/apps/smsforms/tasks.py#L56-L79

What I will do though is update the code linked above to apply the same error status to the subevent (2e65d67db37) but I'm going to leave the SMS code as a fallback.

orangejenny replied:
Is there value to keeping the two different error constants? They seem like they mean the same thing, just in different workflows.

snopoke replied:
I think there is value in keeping them separate since they are on different models and the UI text would be different (more generic for the SMS error than for the subevent error.

I actually wonder if I shouldn't create a new message for this one since the UI text for the other one doesn't really make sense in the context of sms surveys: "No message available for the given language settings."

orangejenny replied:
Ah, I hadn't looked up the message text for `ERROR_NO_MESSAGE`, I was just going based on the constant name.

mkangia commented on corehq/apps/smsforms/tasks.py:
```diff
@@ -59,6 +59,7 @@ def send_first_message(domain, recipient, phone_entry_or_number, session, respon
         metadata = MessageMetadata(
             workflow=workflow,
             xforms_session_couch_id=session.couch_id,
+            messaging_subevent_id=logged_subevent.pk
```
Great!

snopoke committed:
update subevent when no message text is present
